aws/tagsAppmesh.go

```diff
@@ -0,0 +1,128 @@
+package aws
+
+import (
+	"log"
+	"regexp"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/appmesh"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+// setTags is a helper to set the tags for a resource. It expects the
+// tags field to be named "tags"
+func setTagsAppmesh(conn *appmesh.AppMesh, d *schema.ResourceData, arn string) error {
+	if d.HasChange("tags") {
+		oraw, nraw := d.GetChange("tags")
+		o := oraw.(map[string]interface{})
+		n := nraw.(map[string]interface{})
+		create, remove := diffTagsAppmesh(tagsFromMapAppmesh(o), tagsFromMapAppmesh(n))
+
+		// Set tags
+		if len(remove) > 0 {
+			input := appmesh.UntagResourceInput{
+				ResourceArn: aws.String(arn),
+				TagKeys:     remove,
+			}
+			log.Printf("[DEBUG] Removing Appmesh tags: %s", input)
+			_, err := conn.UntagResource(&input)
+			if err != nil {
+				return err
+			}
+		}
+		if len(create) > 0 {
+			input := appmesh.TagResourceInput{
+				ResourceArn: aws.String(arn),
+				Tags:        create,
+			}
+			log.Printf("[DEBUG] Adding Appmesh tags: %s", input)
+			_, err := conn.TagResource(&input)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// diffTags takes our tags locally and the ones remotely and returns
+// the set of tags that must be created, and the set of tags that must
+// be destroyed.
+func diffTagsAppmesh(oldTags, newTags []*appmesh.TagRef) ([]*appmesh.TagRef, []*string) {
+	// First, we're creating everything we have
+	create := make(map[string]interface{})
+	for _, t := range newTags {
+		create[*t.Key] = *t.Value
+	}
+
+	// Build the list of what to remove
+	var remove []*string
```

Comment:
Need to handle the overlapping tags case, e.g. from ECR https://github.com/terraform-providers/terraform-provider-aws/blob/1bb0dead12829bf6c9da9e3170f52a8f24984497/aws/tagsECR.go#L79-L89

See https://github.com/terraform-providers/terraform-provider-aws/issues/7052.

Reply:
Fixed it.